10_cdev/chardev.c

```diff
@@ -0,0 +1,366 @@
+static struct proc_dir_entry *proc_file;
+static struct proc_dir_entry *proc_file2;
+module_param(length, int, 0);
+
+static int     dev_open(struct inode *, struct file *);
+static int     dev_release(struct inode *, struct file *);
+static ssize_t dev_read(struct file *, char *, size_t, loff_t *);
+static ssize_t dev_write(struct file *, const char *, size_t, loff_t *);
+static int volume_read(struct file *file_p, char __user *buffer, size_t length,
+		     loff_t *offset);
+static int size_read(struct file *file_p, char __user *buffer, size_t length,
+		     loff_t *offset);
+
+static struct file_operations volume_fops = {
+	.read = volume_read,
+};
+
+static struct file_operations size_fops = {
+	.read = size_read,
+};
+
+static struct file_operations fops = {
+	.open = dev_open,
+	.read = dev_read,
+	.write = dev_write,
+	.release = dev_release,
+};
+
+static ssize_t sys_show(struct class *class, struct class_attribute *attr,
+			char *buf)
+{
+	int len;
+
+	len = sprintf(buf, "%d", numberOpens);
+	if (len == 0) {
+		pr_err("Failed to convert int to string\n");
+		return 0;
+	}
+	pr_info("%s", buf);
+	pr_info("read %ld\n", (long)strlen(buf));
+	return strlen(buf);
+}
+
+static ssize_t clear_buf(struct class *class, struct class_attribute *attr,
+			const char *buf, size_t count)
+{
+	size_of_message = 0;
+	return count;
+}
+
+static int create_buffer(void)
+{
+	size_buffer = kmalloc(PROC_SYS_BUF, GFP_KERNEL);
+	if (size_buffer == NULL)
+		return -ENOMEM;
+
+	volume_buffer = kmalloc(PROC_SYS_BUF, GFP_KERNEL);
+	if (volume_buffer == NULL)
+		return -ENOMEM;
+
+	return 0;
+}
+
+static void cleanup_buffer(char *buff)
+{
+	kfree(buff);
+	buff = NULL;
+}
+
+static int create_proc_dir(void)
+{
+	proc_dir = proc_mkdir(PROC_DIRECTORY, NULL);
+	if (proc_dir == NULL)
+		return -EFAULT;
+
+	proc_file = proc_create(PROC_FILENAME_1, S_IFREG | 0444, proc_dir,
+				&volume_fops);
+	if (proc_file == NULL)
+		return -EFAULT;
+
+	proc_file2 = proc_create(PROC_FILENAME_2, S_IFREG | 0444, proc_dir,
+				&size_fops);
+	if (proc_file == NULL)
+		return -EFAULT;
+
+	return 0;
+}
+static int size_read(struct file *file_p, char __user *buffer, size_t _length,
+		     loff_t *offset)
+{
+	size_t left;
+	int len;
+
+	if (*offset > 0)
+		return 0;
+
+	len = sprintf(size_buffer, "%d", size_of_message);
+	if (len == 0) {
+		pr_err("Failed to convert int to string\n");
+		return 0;
+	}
+
+	left = copy_to_user(buffer, size_buffer, len);
+	if (left)
+		pr_err("Failed to read %u from %u chars\n", left,
+			length);
+	else
+		pr_notice("Read %u chars, %s\n", len, buffer);
+
+	return len;
+}
+
+static int volume_read(struct file *file_p, char __user *buffer, size_t _length,
+		     loff_t *offset)
+{
+	size_t left;
+	int len;
+
+	if (*offset > 0)
+		return 0;
+
+	len = sprintf(volume_buffer, "%d", length);
+	if (len == 0) {
+		pr_err("Failed to convert int to string\n");
+		return 0;
+	}
+
+	left = copy_to_user(buffer, volume_buffer, len);
```

Comment:
This line contains mistake that lead to garbage in function output.
Please read documentation on sprintf and copy_to_user functions and then fix the mistake.
![image](https://user-images.githubusercontent.com/79188607/112766788-09a19300-901c-11eb-952d-babfbe5e66ce.png)